Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -265,434 +237,255 @@ final class ConnectionPool {
-                    }
-
-                    if connection.channel.eventLoop === channelEL {
-                        return .succeed(firstWaiter.promise)
-                    } else {
-                        if requiresSpecifiedEL {
-                            connection.mustRunDefaultCloseCallback = false
-                            return .replaceConnection(channelEL, firstWaiter.promise)
-                        } else {
-                            return .makeConnectionAndComplete(channelEL, firstWaiter.promise)
-                        }
-                    }
-
-                } else {
-                    connection.isLeased = false
-                    self.leased -= 1
-                    if connection.isActiveEstimation, !connection.isClosing {
-                        self.availableConnections.append(connection)
-                    }
+            self.execute(action)
+        }
 
-                    if self.providerMustClose() {
-                        self.removeFromPool()
-                    }
+        private func makeConnection(on eventLoop: EventLoop) -> EventLoopFuture<Connection> {
+            let handshakePromise = eventLoop.makePromise(of: Void.self)
+            let bootstrap = ClientBootstrap.makeHTTPClientBootstrapBase(group: eventLoop, host: self.key.host, port: self.key.port, configuration: self.configuration)
+            let address = HTTPClient.resolveAddress(host: self.key.host, port: self.key.port, proxy: self.configuration.proxy)
 
-                    return .none
-                }
+            let channel: EventLoopFuture<Channel>
+            switch self.key.scheme {
+            case .http, .https:
+                channel = bootstrap.connect(host: address.host, port: address.port)
+            case .unix:
+                channel = bootstrap.connect(unixDomainSocketPath: self.key.unixPath)
             }
 
-            fileprivate mutating func removeClosedConnection(_ connection: Connection) -> ClosedConnectionRemoveAction {
-                if connection.isLeased {
-                    if let firstWaiter = self.waiters.popFirst() {
-                        let (el, _) = self.resolvePreference(firstWaiter.preference)
-                        return .makeConnectionAndComplete(el, firstWaiter.promise)
+            return channel.flatMap { channel -> EventLoopFuture<Connection> in
+                channel.pipeline.addSSLHandlerIfNeeded(for: self.key, tlsConfiguration: self.configuration.tlsConfiguration, handshakePromise: handshakePromise)
+                return handshakePromise.futureResult.flatMap {
+                    channel.pipeline.addHTTPClientHandlers(leftOverBytesStrategy: .forwardBytes)
+                }.map {
+                    let connection = Connection(key: self.key, channel: channel, pool: self.pool)
+
+                    channel.closeFuture.whenComplete { _ in
+                        self.delete(connection: connection)
                     }
```

Comment:
I don’t think this is enough. We might never call `delete` if we fail before setting up this close handler. Basically if we skip the whole `map` because the future is failed already.

We need a test for this. A connection that is closed before this runs should do it

Reply:
not sure I understand, if connection is closed before this we do not create the connection at all and there is no need to delete it. This callback is only needed to clean up the `Connection`, not the `Channel`

Reply:
@artemredkin what if it fails between `channel.flatMap` and us adding `channel.closeFuture.whenComplete`? So if adding the SSL handler fails or so? Is there no cleanup to do then?

Reply:
I think no. There are two cleanup steps we need to do if connection bootstap fails:
1. decrease `openConnectionCount` - is executed always in `flatMapError`, so it will be called in the scenario you describe
2. delete the `Connection` object, but this is only needed if it was added.
Does this sound reasonable?

Reply:
sounds good, yes